### ISSUE TITLE
M3: one scanned paper end to end

### ISSUE THREAD

**Issue by tamnd** [tamnd/papers-reader#4]
The second paper is deliberately the opposite shape from the first. `codd-1970-relational` is CACM 1970, a scan, two columns, with tables, no figures worth cropping, and a bibliography in a style the parser has not seen.

This is the milestone that proves the vision path rather than the layout path.

### Acceptance

The same audit is clean over both papers, and `reports/usage.md` has real numbers in it for the vision stage.

### Checklist

- [x] `papers render` rasterises pages, which is pure CPU and can run ahead of everything else
- [x] The vision extraction path, with the page image cached and the prompt pinned
- [x] Two column reading order, which is where a naive reader produces fluent nonsense
- [x] Table extraction, and an honest fallback when a table will not survive the trip to Markdown
- [x] `papers pagemap` records which crop came from which page and box
- [x] A second bibliography label style in the reference parser
- [x] Acceptance rules for a page that came through a model, distinct from the rules for a page that did not
- [x] `reports/usage.md` counting tokens, pages and money per stage

**Comment by tamnd** [comment]
#26 is merged, so the first item is done.

`papers render` turns pages into PNGs with pdftoppm and keeps them under
`images/<id>/<profile>/NNNN.png`, one directory per profile so that a page
rendered at 300 dpi in grey and the same page at 600 in colour are two files and
not one file that changed meaning. A page that is already rendered is not
rendered again, which is what makes the command free to re-run and lets it run
days ahead of the stage that consumes it.

The profile is the resolution and whether the page is grey. `pdfimages -list`
says which pages have a colour image on them, and the rest go to grey, which is
about a third of the bytes for the same reading. The ladder is 300, 400 and 600
dpi, and it is one list in the code so that the rasteriser and the retry in the
reader cannot disagree about what the rungs are.

The run over the corpus did 165 pages, found 3 already there, and skipped 2
papers that are not fetched. Nothing it wrote is committed: `images/` is
gitignored, like `pdf/` and the corpus `work/`, and audit rule S03 checks git's
index rather than trusting the ignore file.

**Comment by tamnd** [comment]
#27 is merged. Two more items done: the vision extraction path, and the
acceptance rules for a page that came through a model.

`papers extract --path vision` renders a page, shows it to a model, checks the
answer and writes it, then starts the next one. Each page is written as it
arrives and the extraction record is rewritten after every page, so an
interrupted run keeps everything it paid for and the record describes pages
that are really on disk.

The prompt is `prompt/ocr_paper.md` and its sha256 goes into the record and
from there into `prompt_sha256` in the front matter of every file the pages
become. Editing the prompt changes the hash, which is what makes the pages read
under an older wording findable later. A paper can have a note of its own in
`prompt/notes/<id>.md`, appended to the shared prompt for that paper only,
because a sentence about one paper's symbols added to the shared prompt would
put every other paper's pages back in the queue for a rule that is about none
of them.

Codd's note is the only one and it was written after reading the three rendered
pages. The draft written from the paper's reputation was wrong twice: the folio
is in the running foot and not the head, and the relational algebra symbols
everybody associates with this paper are nowhere in it, because Codd had not
invented them yet. The note as it stands also carries three things a guess
would have missed: the run-in heads at 1.2.1 to 1.2.3, which are not headings,
Structures 1 to 5, which are indentation rather than a grid, and Figures 1 and
2, which are arrays of data and so are the one place the rule about not
transcribing what is inside a figure does not apply.

On the rules, the two that behave differently for a model are A5 and A6. A5 is
about how long this paper's pages usually are, so it stays off for pdftotext,
which cannot truncate a page. A6 needs the page map, and the map is learned
from the folios the pages print. `LearnMap` wants every page at once, which the
vision path cannot offer because its pages arrive over hours, so
`extract.Folios` learns the offset as the pages are accepted and applies it to
the pages after them. The first two pages go unchecked by A6, which is the
price, and the rule still catches the page where the reader lost its place,
because that page disagrees with its neighbours wherever it fell in the run.

`extract.Tidy` takes the wrapping off an answer before the rules see it, since
a model asked for Markdown often returns it in a fence or under a sentence
introducing it. It is narrow on purpose. Its own test caught it eating "The
following is proved in Section 3:", which is a sentence of a paper and not an
introduction to an answer.

A refused page is asked again at 400 dpi and then at 600, which is
`render.Ladder`, the same list the rasteriser uses. This is the only retry in
the toolchain, and a page refused at all three is reported and left alone
rather than failing the run.

Nothing has been extracted yet. There is no route file on this machine, so the
path has nothing to call and says so:

```
  codd-1970-relational    no route can take a page image (the routing table is
  built-in): add one with vision set, or run papers routes init
```

`reports/usage.md` gets real numbers when a real run has happened. Codd's paper
is `access: restricted` either way, so what it can ever publish is its front
matter and an abstract under 250 words, and the pages stay under `work/` and
`images/`, which are gitignored.

**Comment by tamnd** [comment]
A second bibliography label style is done, in #28.

The number style itself has been in the parser since M2: a bibliography whose entries are labelled `1.`, `2.`, `3.` down the left margin, against the bracket style that the modern preprints use. What this was really about was auditing it against a paper that is actually set that way, and the 1970 Communications of the ACM shape found two defects that no fixture had caught.

The first is that the parser reads the word between two author names and only knew it in lower case, so a bibliography that sets its authors in small capitals came out with an author called "R. Q. AND DUNNE". The second is that the older journals put the page range at the end of the entry with nothing in front of it, as in "Comm. ACM 12, 9 (Sept. 1969), 501-507", where the modern styles write "pp." or "pages". The parser was finding no pages at all for those entries, which means every one of them was going into the corpus missing the one field a reader uses to go and find the paper.

Both are fixed and both have a test. The third test in that PR is the whole shape, a numbered bibliography in the older style, read end to end.

**Comment by tamnd** [comment]
Tables are done, in #29, and two column reading order is done, in #31. They came out of the same session because auditing the first found the second.

Tables are found by geometry, anchored to a caption. Words go back into the rows they were printed as before anything looks for a grid, because pdftotext emits a line per run of text set in one size and the ablation table of the Transformer paper arrives one cell at a time, forty pieces for nine rows. A run of rows cut into cells the same way is a table if a caption stands within five rows above it or three below.

The caption is required, and so is a guard on how many words a cell holds. Between them they are the answer to the thing that worried me most about this item: a two column page whose gutter was missed is a perfect grid, and reading it as a table would be worse than the interleaved sentences it started as. A table's cells are not sentences, and a half line of a two column page runs to nine or ten words, so a run whose middle cell holds more than six is not a table.

Two of the four tables in the Transformer paper have a cell standing over two columns, which cannot be a pipe table. Dropping the cell, or moving it into one of the columns it stands over, would make the corpus say something the paper does not, so those come out inside a fence at the layout the page laid them out in. That is the honest fallback this box asked for.

The two column item was already mostly built, in #24, and I was going to tick it on the strength of that until I read the real output. The last body row of the first page of the Amdahl paper was published as three columns interleaved in one line, and the paragraph above it stopped halfway through a sentence with the rest of it nowhere on the page. The cut at the gutter has two tests and the weaker one was being measured against the row's own word spacing, which for a row that already runs across every column is three columns' worth of justification plus the gutter steps themselves. That came out at eight points against a page whose word space is five, which lifted the threshold just over the gutter. It is now a quarter of the way up the row's gaps instead of the middle of them.

Both have tests and every fixture in them is typeset in the test.

One thing I want to record rather than hide. I extracted `codd-1970-relational` down the native path to check the two column order on a paper that is actually set in two columns, and it is not good enough to publish: the text layer is OCR, and the two figures on its second page come out interleaved with the prose beside them. That is the paper this milestone says proves the vision path, so it stays unextracted until the vision path can run against a real model here. It has been backed out of the corpus rather than committed.

**Comment by tamnd** [comment]
`papers pagemap` is done, in #33, and it found a defect on its first run over the corpus which is fixed in #32.

The command writes `manifests/pages/<id>.yaml`: the page of the file against the page the paper prints, the size of each page, how many columns it was set in, and the figures cut from it. The offset at the top of the file is learned from the folios, by the same median rule and the same code acceptance rule A6 uses, so a paper cannot have one page map during extraction and another one in its manifest.

The page size is there for the reason this checklist item names. A bounding box in `figures.yaml` is four numbers in points, and four numbers in points mean nothing without the page they were measured on: a box 300 points wide is half of a letter page and a fifth of a poster. The box itself stays in `figures.yaml` rather than being copied, because two manifests holding one number is two manifests that can disagree. The figure list per page is the same fact from the other end, which is the question somebody actually asks when a figure looks wrong.

The defect. A bare number in the margin was being read as the page number the page printed. A two column paper sets its footnotes at the foot of the columns, and a footnote marker is a bare number in the bottom margin, so pages 3 to 6 of the BERT paper read as printing 4, 6, 8 and 10. That paper prints no page numbers anywhere, and the map learned from those four markers put its first page at page 3 of a paper with no page 3. That is worse than a wrong manifest: A6 refuses a page whose printed number disagrees with the map, so a paper with a wrong map can refuse good pages.

What separates a folio from a footnote marker is that a folio counts with the file. Every margin number is a candidate now, the offsets are counted, and the offset a third of the pages agree on wins. Two agreeing is not enough, which the first version of the fix got wrong: a footnote on page nine and the next footnote on page ten are a page apart and look exactly like a folio.

98 papers map. Three are skipped, one unfetched and two with nothing recorded about what may be published from them. A second run writes nothing.

That leaves one box on this milestone, `reports/usage.md` with real numbers for the vision stage, and that one is blocked on there being a route file on this machine. I would rather say so than write a report with invented numbers in it.

**Comment by tamnd** [comment]
`reports/usage.md` is written and committed, in #34, and that is the last box on the checklist. The milestone is not done, and here is exactly what is left.

`papers report usage` rolls the ledger up into `reports/usage.md`: asks, pages, tokens, waiting and money, by stage, by model and by paper. The ledger is one line per ask and lives beside the routing table under `~/.config/papers`, because it names the hosts that were asked. The report is committed to a public repository and names none of them, and it does not print the resolved path of the ledger either, because a resolved path has a home directory in it and a home directory has a user name in it. There is a test that fails if a host name ever turns up in the markdown.

What the report says today is that nothing has asked a model anything on this machine. Every stage row is a zero, and the report says so in its own words rather than printing nothing. That is honest and it is not the acceptance:

> The same audit is clean over both papers, and `reports/usage.md` has real numbers in it for the vision stage.

The vision stage has never run here. There is no `~/.config/papers/routes.json`, so `papers extract -path vision` has nowhere to send a page, and `codd-1970-relational` is still the paper the milestone is named after and still unextracted. So the milestone stays open on that one thing, and I would rather leave it open than tick it with a page of invented numbers.

What the report does count today, off the committed content rather than off the ledger, is the pages:

| path | papers | pages | what read them |
| --- | --: | --: | --- |
| native | 6 | 32 | pdftotext version 26.09.0 |

That section exists because a report of what the work cost that counted only the asks would leave out every page that cost nothing, and those are most of them and the ones whose text was never guessed.

Two other things worth writing down. The money column is a dash for a model that is not in a price table, and a stage that mixes a priced model with an unpriced one is a dash as well, because the sum of the half that was priced looks exactly like the whole. And `papers report usage -write` refuses to write a report of nothing over a report with a night of work in it: the ledger is on the machine that did the work and the report is in the corpus, so the two come apart on a second checkout, and overwriting there would look like a stage that had been reverted rather than a report rebuilt from nothing.

Released as v0.3.2, which also carries #32 and #33. The corpus CI pin moved with it.

**Comment by tamnd** [comment]
M3 is done. Both papers are read, the audit is clean over them, and `reports/usage.md` has real numbers in it for the vision stage, which is what the acceptance asked for.

The vision path ran for the first time this week, against a local GPU serving olmOCR-2-7B-1025-FP8 and against the gateways as a fallback. Everything below came out of watching what it actually returned, which was not what the prompt asked for in four separate ways.

The reader answers in TeX's own delimiters whatever the prompt says, so `\(` and `\[` are read as dollars rather than refused (#37). It writes a Markdown image link for a figure that was never rendered, and four of those were hard F01 failures, so an invented link is dropped (#39). It answers with `<table>` because that is what its training data looks like, so an HTML table is converted to a pipe table (#43) and a rule was added for the ones that are left (#43 again, T11).

Then #44, which is the one worth reading twice. The first cut of the table converter produced a wrong Table 3 for the Transformer paper and did it silently. The reader had written `rowspan="4"` over a group of five rows and `colspan="7"` to mean seven empty columns rather than one cell seven wide, and expanded as written every number after the first group landed under the wrong heading. A table whose numbers are under the wrong headings is worse than no table, because it reads like a table. So a table is only converted when its cells form a rectangle: every row the same width after the spans are expanded, and no span hanging over the end. Table 2 of the same paper uses both kinds of span correctly and converts. Table 3 stays HTML and T11 reports it hard, which is the right answer.

#41 moved the acceptance rules off the page and onto the paper. A page of 300 characters is a refusal in a paper that averages 3500 and a normal page in a paper that averages 400, and judging it against the pages either side of it made the first and last page of every paper suspicious.

#45 is the hand edit deadlock. The splitter protected a correction by noticing `content_sha256` no longer matched the body, and rule T03 failed a file whose `content_sha256` no longer matched the body, so a corrected corpus could never be green. `papers split --accept` restamps the hash and writes `edited: true`, the splitter protects the file by the field rather than by the mismatch, an unaccepted edit is still protected, and `--force` still overrides both.

#44 also carries `papers extract --retidy`. When `Untable` was added, four pages of the Transformer paper were already on disk with their tables in HTML, and the choice was between re-reading them on a rationed model that might come back worse and running the new step over text that was already right. Every step in the tidier is a translation between two spellings, so running it over a tidy page writes nothing, and there is a test that fails the day that stops being true.

Then `codd-1970-relational` itself, which is the paper this milestone is named after. It is restricted, so the whole of what may ever be published from it is the front matter and an abstract, and three pages were read and the rest of the file was not. Those three pages came back clean, and reading them found the last defect: #46. The reader left a blank line under the first of its four headings and none under the other three, so only one of the four was a paragraph of its own. The numbering chain needs three in sequence before a scheme is believed, so the scheme came out as none, the numbered pass found nothing, the typography fallback turned on, and a subheading set in capitals was published as the only section of a paper that numbers its sections perfectly well. Nothing was refused and nothing was reported. The splitter now separates a heading from the paragraph it was run into before anything looks for headings.

Where that leaves the corpus: two papers in full, five as front matter and an abstract, the audit clean at 64 rules and 64 passed, and a usage report that says 52 asks, 50 answered, 564,131 tokens in, 31,553 out and 35 minutes of waiting over 27 pages. The pages read table now has a vision row against the native one.

Released as v0.4.0, the mirror release for this milestone. The corpus CI pin moved with it.

Two things found here that are not this milestone's to fix and have issues of their own coming: acceptance rule A5 refuses a figure-only page, which is correct arithmetic and the wrong answer for a page that is mostly picture, and the gateway route truncates an answer to a couple of hundred characters often enough to be worth measuring.
